File: cpp/0063-unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obs) {
        int m = obs.size(), n = obs[0].size();
        if (obs[0][0] == 1) return 0;
        vector<int> dp(n, 0);
        dp[0] = 1;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (obs[i][j] == 1) dp[j] = 0;
                else if (j > 0) dp[j] += dp[j - 1];
            }
        }
        return dp[n - 1];
    }
};

File: python/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obs):
        m, n = len(obs), len(obs[0])
        if obs[0][0] == 1:
            return 0
        dp = [0] * n
        dp[0] = 1
        for i in range(m):
            for j in range(n):
                if obs[i][j] == 1:
                    dp[j] = 0
                elif j > 0:
                    dp[j] += dp[j - 1]
        return dp[-1]
